Sources/Valkey/Subscriptions/SubscriptionConnectionManager.swift

```diff
@@ -0,0 +1,140 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the valkey-swift project
+//
+// Copyright (c) 2025 the valkey-swift authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See valkey-swift/CONTRIBUTORS.txt for the list of valkey-swift authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Synchronization
+
+/// Stores a reference to a single connection to be used by client for subscriptions
+///
+/// It ensures only ever one version of the connection is initialized, even if it requested
+/// twice during initialization. Once it is available the object includes a reference count
+/// so we can clean it up once nobody references it.
+@available(valkeySwift 1.0, *)
+@usableFromInline
+package struct SubscriptionConnectionManager: ~Copyable, Sendable {
+    @usableFromInline
+    enum Action {
+        case use(ValkeyConnection)
+        case acquire
+    }
+
+    @usableFromInline
+    enum State {
+        case uninitialized
+        case acquiring([CheckedContinuation<Action, any Error>])
+        case available(ValkeyConnection, Int)
+    }
+    @usableFromInline
+    let state: Mutex<State>
+
+    init() {
+        self.state = .init(.uninitialized)
+    }
+
+    @usableFromInline
+    package func acquire(
+        isolation: isolated (any Actor)? = #isolation,
```

Comment:
I think we should run the `acquire` action inside the client's background task, otherwise we run into issues with cancellation again.